common/enclave/common.cpp

```diff
@@ -49,20 +65,26 @@ int ecall_init(void)
 
     LOG_DEBUG("Enc: Identity generated!");
 
-    {  // TODO: this block is here temporarily, only meant to test the attestation api
-        char params[] = "{\"attestation_type\":\"simulated\"}";
-        // char params[] = "{\"attestation_type\":\"simulated\"}";
-        bool bb = init_attestation((uint8_t*)params, strlen(params));
-        COND2LOGERR(bb == false, "error init attestation");
-        LOG_DEBUG("init attestation successful");
-
-        char statement[] = "1234567890";
-        uint8_t attestation[2048];
-        uint32_t attestation_length = 0;
-        bool b = get_attestation(
-            (uint8_t*)statement, strlen(statement), attestation, 2048, &attestation_length);
-        COND2LOGERR(b == false, "error getting attestation");
-        LOG_DEBUG("get attestation success, size %d", attestation_length);
+    // NOTE:
+    // cc_data is a global pointer, meant to reference a global variable of cc_data type.
+    // If g_cc_data is implemented as a simple global variable, cgo seems to crash (on enclave
```

Comment:
It's "C"Go, not "C++"Go.
The problem lies with the class constructors/destructors when you put global variables. CGo apparently does not handle these properly. So, the workaround is to use a pointer (and dynamically allocate the object) rather than a simple variable.

Reply:
but we do not have any cgo code inside the enclave, this is all outside; why should that affect any code running inside the enclave?

Reply:
We start from Go, and have Go <--> CGo <--> C <--> Cpp.
Everything works fine with function calls, local/global variables of basic types (e.g., int), and local objects.
Global objects however require constructors/destructors to be called, and this creates issues.
In particular, if you implement `g_cc_data` as a global object, ECC crashes.

Reply:
I'm not doubting that ecc crashes with a static g_cc_data object, just having problem to see that cgo is the real culprit. I.e., the chain is really  'Go <--> CGo <--> C <-> ecall/ocall <-> C <--> Cpp', so while maybe cgo could influence input parameters for an ecall, it shouldn't (be able to) affect any static variable/initialization inside the enclave? That seems more to indicate that maybe some ecall was not called or any input (or output) from an ecall (ocall) triggered a bug inside the enclave where some input-validation was missing/incorrect?

Reply:
> it shouldn't (be able to) affect any static variable/initialization inside the enclave?

In fact it doesn't. Again, the problem is the constructor/destructor call (I think the latter).

> a bug inside the enclave where some input-validation was missing/incorrect?

Again, given that dynamically allocation works, while static one does not, this does not seem the case.

> That seems more to indicate that maybe some ecall was not called

Right, at init or exit time.
This issue [here](https://github.com/bazelbuild/rules_go/issues/1486) might be related.

Reply:
but this is linktime issue of the untrusted part of the app, this shouldn't affect the linking of enclave.so (where your static variable resides i though)?

Reply:
anyway, it should work as is and i don't even see a situation where we would have to free it.    I guess my main reason to follow-up was curiousity and the thought that this might have unearthed another latent issue.  But if that is the case, we can postpone it to the time it re-appears (if at all ...)